### PR TITLE
Allow to selectively disable timing

### PR DESCRIPTION
**Pull request by sloede**
Relevant discussion on Discourse: https://discourse.julialang.org/t/disable-timeroutputs-jl-timers-selectively-within-function/35957

### CODE REVIEW

**Comment by sloede** [comment]
Hi @KristofferC, any chance you find the time to have a quick look at this (= my comment above)? We have encountered this problem more often recently, and it would be a great help to have this functionality (especially for nested function calls).

File: src/TimerOutputs.jl [hunk 1]
```diff
@@ -1,7 +1,7 @@
 module TimerOutputs
 
 import Base: show, time_ns
-export TimerOutput, @timeit, @timeit_debug, reset_timer!, print_timer, timeit
+export TimerOutput, @timeit, @timeit_debug, reset_timer!, print_timer, timeit, enable!, disable!
```
Comment (fredrikekre) [inline src/TimerOutputs.jl:4]:
```suggestion
export TimerOutput, @timeit, @timeit_debug, reset_timer!, print_timer, timeit, enable_timer!, disable_timer!
```
perhaps? Fits better with the other exported functions.

**Comment by KristofferC** [comment]
I added some changes here. Please try them out.

**Comment by sloede** [comment]
Seems to work well, thanks! One suggestion:

From a usability point of view, IMHO it would make sense to add a macro that allows to selectively disable a timer for a given expression, similar to `@timeit`:
```julia
const to = TimerOutput()
fun() = @timeit to "fun" 1+1

# This time we record "fun"
fun()

# This time we don't
@notimeit to fun()
```
To be conistent with `@timeit`, it should also support to omit the `to` argument and use the default timer in this case instead.

Do you think this would be useful? If not, from my point of view this PR looks good to be merged.

**Comment by KristofferC** [comment]
I don't really understand that. What is the difference between `@notimeit to fun()` and just `fun()`?

**Comment by sloede** [comment]
Sorry for not being clear enough:
```julia
@notimeit to fun()
```
should be functionally equivalent to
```julia
disable_timer!(to)
fun()
enable_timer!(to)
```

The use case I had in mind was as follows (based on our personal experience): We use a single timer to time different parts of a numerical simulation framework, and we rely on `TimerOutput`'s automatic nesting capabilities to accumulate and present timing data in well-arranged manner after each simulation. However, sometimes functions that include timed regions (blocks, or other function calls) are called 'out-of-order', i.e., from a call site where we are not within the regular timing hierarchy. In this case, we usually do not want the timer to be active, as it clutters output of `print_timer`. Therefore it would be handy if we have a simple macro call to disable a given timer within a certain block/function and re-enable it afterwards. Does that makes sense?

**Comment by sloede** [comment]
I took the liberty to implement the `@notimeit` macro as I had envisioned it. Feel free to comment (and please check if I got the implementation of `notimeit_expr` right :grimacing: ).

File: src/TimerOutput.jl [hunk 2]
```diff
@@ -311,3 +311,29 @@ end
...
+# If @notimeit was called without a TimerOutput instance, use default timer
+notimeit_expr(ex::Expr) = notimeit_expr(:($(TimerOutputs.DEFAULT_TIMER)), ex)
+
+# Disable timer, evaluate expression, restore timer to previous value, and return expression result
+function notimeit_expr(to, ex::Expr)
+  return quote
+    local to = $(esc(to))
+    local enabled = to.enabled
+    $(disable_timer!)(to)
+    local val = $(esc(ex))
```
Comment (KristofferC) [inline src/TimerOutput.jl:333]: I wonder if you need a `try finally` here in case the expression throws?

**Comment by sloede** [comment]
Thanks for the comment. I added the try-finally based on the implementation used in `timer_expr` and also fixed the tests.

**Pull request merged**